src/translators/sql_translator.py

```diff
@@ -0,0 +1,964 @@
+        if fiware_service:
+            where = "where table_name ~* '\"{}{}\"[.].*'"
+            op += where.format(TENANT_PREFIX, fiware_service.lower())
+
+        try:
+            self.cursor.execute(op)
+        except Exception as e:
+            # Metadata table still not created
+            msg = "Could not retrieve METADATA_TABLE. Empty database maybe?. {}"
+            logging.debug(msg.format(e))
+            return []
+        return [r[0] for r in self.cursor.rows]
+
+
+    def _get_select_clause(self, attr_names, aggr_method, aggr_period):
+        if not attr_names:
+            return '*'
+
+        attrs = ['entity_type', 'entity_id']
+        if aggr_method:
+            if aggr_period:
+                attrs.append(
+                    "DATE_TRUNC('{}',{}) as {}".format(
+                        aggr_period, self.TIME_INDEX_NAME, self.TIME_INDEX_NAME)
+                )
+            # TODO: https://github.com/smartsdk/ngsi-timeseries-api/issues/106
+            m = '{}("{}") as "{}"'
+            attrs.extend(m.format(aggr_method, a, a) for a in set(attr_names))
+
+        else:
+            attrs.append(self.TIME_INDEX_NAME)
+            attrs.extend('"{}"'.format(a) for a in attr_names)
+
+        select = ','.join(attrs)
+        return select
+
+
+    def _get_limit(self, limit, last_n):
+        # https://crate.io/docs/crate/reference/en/latest/general/dql/selects.html#limits
+        default_limit = 10000
+
+        if limit is None or limit > default_limit:
+            limit = default_limit
+
+        if last_n is None:
+            last_n = limit
+
+        if limit < 1 or last_n < 1:
+            raise NGSIUsageError("Limit and LastN should be >=1 and <= {"
+                                 "}.".format(default_limit))
+
+        return min(last_n, limit)
+
+
+    def _get_where_clause(self, entity_ids, from_date, to_date, fiware_sp=None,
+                          geo_query=None):
+        clauses = []
+
+        if entity_ids:
+            ids = ",".join("'{}'".format(e) for e in entity_ids)
+            clauses.append(" entity_id in ({}) ".format(ids))
+        if from_date:
+            clauses.append(" {} >= '{}'".format(self.TIME_INDEX_NAME,
+                                                self._parse_date(from_date)))
+        if to_date:
+            clauses.append(" {} <= '{}'".format(self.TIME_INDEX_NAME, self._parse_date(to_date)))
+
+        if fiware_sp:
+            # Match prefix of fiware service path
+            clauses.append(" "+FIWARE_SERVICEPATH+" ~* '"+fiware_sp+"($|/.*)'")
+        else:
+            # Match prefix of fiware service path
+            clauses.append(" "+FIWARE_SERVICEPATH+" = ''")
+
+        geo_clause = self._get_geo_clause(geo_query)
+        if geo_clause:
+            clauses.append(geo_clause)
+
+        where_clause = "where " + "and ".join(clauses)
+        return where_clause
+
+    def _parse_date(self, date):
+        try:
```

Comment:
in this way we should have tackled sql injection for dates

Reply:
awesome